doc/developer/design/20210811_catalog_only_mode.md

```diff
@@ -0,0 +1,257 @@
+- This doc does not consider the potential design of disabling indexes at run
+  time, e.g. `ALTER INDEX ... SET DISABLED`, though it seems possible to use a
+  process similar to booting without user indexes to remove all dataflows that
+  use the index, and then remove it from the `Catalog`'s `indexes` field.
+
+## Description
+
+There are two main considerations in implementing this mode:
+- Booting without user indexes
+- Enabling user indexes at runtime.
+
+### Booting without user indexes
+
+@mjibson came up with the idea of making this mode behave akin to removing the
+`MATERIALIZED` keyword from all catalog objects before booting and letting the
+system simply behave as it would in that case. (The exception to this behavior
+being tables, which we'll address separately.)
+
+To accomplish this we can propagate the presence of the `--disable-user-indexes`
+flag to the `Coordinator` and its `Catalog`.
+
+The flag's presence on the...
+- `Catalog` will prevent indexes on user-created objects from being added to its
+  `indexes` field (i.e. bypass [this
+  code](https://github.com/MaterializeInc/materialize/blob/8fad35548028ef9eac6866fdf679a57dd90e47c2/src/coord/src/catalog.rs#L1252-L1256)),
+  which is used throughout the `Coordinator` for determining which indexes are
+  available. Not adding them to this field leaves their catalog entries intact
+  for either removal or for use on reboot, but prevents the system from being
+  aware that they exist.
+- `Coordinator` will prevent shipping any dataflows on user indexes, by slightly
+  amending
+  [`build_index_dataflow`](https://github.com/MaterializeInc/materialize/blob/93466416ba9813ec8ff4e52b039d76036ef36329/src/coord/src/coord/dataflow_builder.rs#L156)
+  to return an `Option<DataflowDesc>` that will be `None` on any `id` that is a
+  `GlobalId::User`.
+
+You can see an rought initial implementation of this approach [on a branch on my
+fork](https://github.com/MaterializeInc/materialize/compare/main...sploiselle:disable-user-indexes?expand=1).
+
+#### Table details
+
+This design still lets users `SELECT` from tables, but never returns any data.
+I'm a little fuzzy on why this would be the case, but in my prototype that is
+the observed behavior.
+
+However, we have to prevent users from inserting data into tables because their
+indexes will not exist. To prevent this, we'll need to return an error from
+`sequence_insert` when in `disable_user_indexes` mode when we see that the table
+has no indexes in the catalog. Checking for indexes in a side channel seems like
+the most reasonable design because there is no time during `sequence_insert`
+where we have visibility into which indexes the dataflow layer plans to use to
+service the `INSERT`.
+
+### Enabling indexes
+
+When we receive a `ALTER INDEX ... SET ENABLED` command from the user, we'll:
+
+1. Ensure the named object is an index and error if not.
+1. Check if `disable_user_indexes` is `true` and return OK if so.
```

Comment:
I think this should actually just be a no-op if the index is already enabled or if the server is not in catalog-only mode.

Reply:
Ah, this was what I meant by return `OK`, unless there is some other special nop return value? Wanted something like this:

```
sean=# create table a (a int);
CREATE TABLE
sean=# alter table a alter column a set default 1;
ALTER TABLE
sean=# alter table a alter column a set default 1;
ALTER TABLE
```

Reply:
Heh, I think there are two errors here. One is that I think the text should read "Check if `disable_user_indexes` is _false_", right? And then the other is that my brain read "return error", not "return OK". So tl;dr I think we are both in agreement that the behavior should be

```
ALTER INDEX a SET ENABLED
ALTER INDEX
```

in any situation where the index is already enabled.

Reply:
Yeah, that should have been `false` in the doc. sgtm